src/libraries/System.IO.Packaging/src/System/IO/Packaging/InterleavedZipPackagePartStream.cs

```diff
@@ -0,0 +1,572 @@
+                {
+                    if (pieceStream.Position != pieceStreamRelativeOffset)
+                    {
+                        pieceStream.Seek(pieceStreamRelativeOffset, SeekOrigin.Begin);
+                    }
+                }
+                else
+                {
+                    pieceStream = _dir.ResetStream(pieceNumber);
+                    SeekUnderlyingPieceStream(pieceStream, pieceStreamRelativeOffset);
+                }
+
+                try
+                {
+                    while (totalBytesRead < buffer.Length)
+                    {
+#if !NETFRAMEWORK && !NETSTANDARD2_0
+                        int numBytesRead = pieceStream.Read(buffer.Slice(totalBytesRead));
+#else
+                        int numBytesRead = pieceStream.Read(
+                            tempInputBuffer,
+                            totalBytesRead,
+                            buffer.Length - totalBytesRead);
+
+                        tempInputBuffer.AsSpan(totalBytesRead, numBytesRead).CopyTo(buffer.Slice(totalBytesRead, numBytesRead));
+#endif
+
+                        // End of the current stream: try to move to the next stream.
+                        if (numBytesRead == 0)
+                        {
+                            if (_dir.IsLastPiece(pieceNumber))
+                                break;
+
+                            ++pieceNumber;
+                            Debug.Assert(_dir.GetStartOffset(pieceNumber) == _currentOffset + totalBytesRead);
+
+                            pieceStream = _dir.GetStream(pieceNumber);
+
+                            //Seek inorder to set the correct pointer for the next piece stream
+                            if (pieceStream.CanSeek)
+                            {
+                                if (pieceStream.Position != 0)
+                                {
+                                    pieceStream.Seek(0, SeekOrigin.Begin);
+                                }
+                                else
+                                {
+                                    pieceStream = _dir.ResetStream(pieceNumber);
+                                }
+                            }
+                        }
+
+                        totalBytesRead += numBytesRead;
+                    }
+                }
+                finally
+                {
+#if NETFRAMEWORK || NETSTANDARD2_0
+                    ArrayPool<byte>.Shared.Return(tempInputBuffer);
+#endif
+                }
+
+                // Advance current position now we know the operation completed successfully.
+                _currentOffset += totalBytesRead;
+            }
+
+            return totalBytesRead;
+        }
+
+        /// <inheritdoc/>
+        public override long Seek(long offset, SeekOrigin origin)
+        {
+            CheckClosed();
+
+            // Check stream capabilities. (Normally, CanSeek will be false only
+            // when the stream is closed.)
```

Comment:
I don't know about this comment. Do we have tests that verify the behavior when handling unseekable streams as the input? For example, a network stream?